src/proto/single/mod.rs

```diff
@@ -206,19 +208,39 @@ impl JobBuilder {
     }
 }
 
+/// Details on a job's failure.
 #[derive(Serialize, Deserialize, Debug, Clone)]
+#[non_exhaustive]
 pub struct Failure {
-    retry_count: usize,
-    failed_at: String,
+    /// [`Number`](Job::retry) of times this job can be retried.
+    pub retry_count: usize,
+
+    /// Number of remaining retry attempts.
+    #[serde(rename = "remaining")]
+    pub retry_remaining: usize,
+
+    /// Last time this job failed.
+    pub failed_at: DateTime<Utc>,
     #[serde(skip_serializing_if = "Option::is_none")]
-    next_at: Option<String>,
+
+    /// When this job will be retried.
+    ///
+    /// This will be `None` if there are no retry
+    /// attempts (see [`Failure::retry_remaining`]) left.
+    pub next_at: Option<DateTime<Utc>>,
     #[serde(skip_serializing_if = "Option::is_none")]
-    message: Option<String>,
+
+    /// Error message, if any.
+    pub message: Option<String>,
     #[serde(skip_serializing_if = "Option::is_none")]
     #[serde(rename = "errtype")]
-    kind: Option<String>,
+
+    /// Error kind, if known.
+    pub kind: Option<String>,
```

Comment:
Is kind actually an arbitrary string? Things like this is why `Failure` wasn't previously `pub`.

Reply:
hmm, but then why would we have `failure` accessor in our public API. 

I understand that breaking changes are unpleasant, but do you think the fact that the `kind` may become "enumerated"  in Faktory (even if this enumeration were not exhaustive)  tomorrow should prevent us from exposing `Failure` in the bindings today? Or you mean more research is needed on whether this can enumerated?

Let me double-check though

Reply:
In our codebase, the kind (`errtype` in Faktory) is always an "unknown" literal, we construct a `Fail` of this kind with `Fail::generic` all the time. 

Also "unknown" is what they use in [go bindings](https://github.com/contribsys/faktory/blob/10ccc2270dc2a1c95c3583f7c291a51b0292bb62/client/client.go#L363). Notice how this is actually hard-coded.

I wish we could enumerate this (with a non-exhaustive enum where `Unknown` would be the only variant for now), but what if there are also other workers (in other languages or not using the `faktory-rs` bindings), how do we unmarshal that? 

___

The more I am thinking about it, the more I am questioning if there is any usefulness in this field at all.
We can not make `kind` public for now at all, and if in the future it is no longer a hard-coded "unknown" literal, we could consider enumerating it, wdyt?

Reply:
Yeah, I like that idea — let's just make that field not `pub` for now!